src/neighborsearch/neighborhood.jl

```diff
@@ -63,8 +76,69 @@ function search(xₒ::AbstractVector, method::NeighborhoodSearch{O,N,T};
         push!(neighbors, ind)
       end
     end
+    method.restrictions && (neighbors = filterneighs(neighbors, method, xₒ))
     neighbors
   else
+    method.restrictions && (inds = filterneighs(inds, method, xₒ))
     inds
   end
 end
+
+function filterneighs(inds::AbstractVector, method::NeighborhoodSearch{O,N,T},
+  xₒ::AbstractVector) where {O,N<:BallNeighborhood,T}
+
+  object   = method.object
+  neigh    = method.neigh
+  rotmat   = neigh.metric isa Mahalanobis ? neigh.metric.qmat : I
+  ometric  = method.ordermetric
+  maxk     = method.maxneighs
+  maxoct   = method.maxperoct
+  maxkey   = method.maxperkey
+  usek     = 0 < maxk < length(inds)
+  useoct   = maxoct > 0
+  usekey   = maxkey[2] > 0
+
+  # get distances and give priority to closest neighbors according to ordermetric
+  dists   = colwise(ometric, xₒ, coordinates(object,inds))
+  sortids = sortperm(dists)
+  inds    = inds[sortids]
+
+  # key and octant constraints
+  keys   = usekey ? unique(view(values(object), inds, maxkey[1])) : nothing
+  dkeys  = usekey ? Dict(zip(keys, 1:length(keys)))               : nothing
+  ctkeys = usekey ? zeros(Int, length(keys))                      : nothing
+  octs   = useoct ? zeros(Int, 8)                                 : nothing
+
+  neighbors = Vector{Int}()
+  ctneighs  = 0
+
+  @inbounds for ind in inds
+    key = usekey ? values(object)[ind,maxkey[1]]                    : nothing
+    oct = useoct ? getoct(rotmat * (coordinates(object,ind) .- xₒ)) : nothing
+
+    usekey && ctkeys[dkeys[key]] >= maxkey[2] && continue
+    useoct && octs[oct] >= maxoct             && continue
+
+    # if valid, add as a neighbor
+    push!(neighbors, ind)
+    ctneighs += 1
+    usekey && (ctkeys[dkeys[key]] += 1)
+    useoct && (octs[oct] += 1)
+
+    # if maxneigh reached, stop
+    usek && ctneighs >= maxk && break
+  end
+
+  neighbors
+end
+
+
+# get octant code of transformed coordinates
+function getoct(coords::AbstractVector)
+  N = size(coords,1)
+  i = (coords .< 0) .+ 1
+
+  octid = N == 2 ? reshape(1:4,(2,2)) : reshape(1:8,(2,2,2))
+  oct   = N == 2 ? octid[i[1],i[2]]   : octid[i[1],i[2],i[3]]
+  oct
+end
```

Comment:
Sorry my lack of understanding here, but would you mind explaining the filtering operations you have in mind in a high-level language, possibly with illustrations? I could provide better feedback that way.

Reply:
Here the sketch of the idea

![text1203-8](https://user-images.githubusercontent.com/8386288/102501267-02054200-405c-11eb-80ce-c8591adcc5f4.png)